packages/backend/src/services/PromoteService/PromoteService.ts

```diff
@@ -0,0 +1,761 @@
+            );
+        }
+
+        const upstreamSpaces = await this.spaceModel.find({
+            projectUuid: upstreamProjectUuid,
+            slug: promotedSpace.slug,
+        });
+        if (upstreamSpaces.length > 1) {
+            throw new AlreadyExistsError(
+                `There are multiple spaces with the same identifier ${promotedSpace.slug}`,
+            );
+        }
+        const chartUuids = dashboard.tiles.reduce<string[]>((acc, tile) => {
+            if (isChartTile(tile) && tile.properties.savedChartUuid) {
+                return [...acc, tile.properties.savedChartUuid];
+            }
+            return acc;
+        }, []);
+        const promotedDashboard: PromotedDashboard = {
+            dashboard,
+            projectUuid: dashboard.projectUuid,
+            chartUuids,
+            space: promotedSpace,
+            access: await this.spaceModel.getUserSpaceAccess(
+                user.userUuid,
+                promotedSpace.uuid,
+            ),
+        };
+        const upstreamSpace =
+            upstreamSpaces.length === 1 ? upstreamSpaces[0] : undefined;
+        const upstreamDashboard = {
+            dashboard:
+                existingUpstreamDashboards.length === 1
+                    ? existingUpstreamDashboards[0]
+                    : undefined,
+            projectUuid: upstreamProjectUuid,
+            space: upstreamSpace,
+            access: upstreamSpace
+                ? await this.spaceModel.getUserSpaceAccess(
+                      user.userUuid,
+                      upstreamSpace.uuid,
+                  )
+                : [],
+        };
+        try {
+            await PromoteService.checkPromoteDashboardPermissions(
+                user,
+                promotedDashboard,
+                upstreamDashboard,
+            );
+
+            // Check permissions for all chart tiles
+
+            const chartPromises = chartUuids.map((chartUuid) =>
+                this.getPromoteCharts(user, upstreamProjectUuid, chartUuid),
+            );
+            const charts = await Promise.all(chartPromises);
+
+            charts.forEach(({ promotedChart, upstreamChart }) =>
+                PromoteService.checkPromoteChartPermissions(
+                    user,
+                    promotedChart,
+                    upstreamChart,
+                    promotedDashboard,
+                ),
+            );
+
+            // at this point, all permisions checks are done, so we can safely promote the dashboard and charts.
+            const updatedDashboard = await this.upsertDashboard(
+                user,
+                promotedDashboard,
+                upstreamDashboard,
+            );
+
+            // Upserting all charts
+            // This should not cause any permission issues, as we have already checked them
+            // But if this fails somehow, we should return a partial success response
```

Comment:
right now I don't see a reason for this to fail, but if we get errors we might want to consider returning a partial response.